c10/core/TensorImpl.cpp

```diff
@@ -29,27 +29,19 @@ const char * const TensorImpl::err_msg_tensor_metadata_change_not_allowed =
     "        x.set_(y)";
 
 at::Tensor& TensorImpl::grad() {
-  if (type_set_.has(TensorTypeId::VariableTensorId)) {
-    if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
-    return autograd_meta_->grad();
-  } else {
-    AT_ERROR("grad is not implemented for Tensor");
-  }
+  if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
+  return autograd_meta_->grad();
 }
 
 const at::Tensor& TensorImpl::grad() const {
-  if (type_set_.has(TensorTypeId::VariableTensorId)) {
-    // Yes, I know this looks really weird.  But I don't really have a choice as
-    // long as this function returns a const reference to Tensor.  I'm not
-    // really sure how I would have designed this API differently, but it
-    // is not so easy to fix right now because the mutable counterpart of
-    // this function must keep working so that "x.grad() = ..." keeps working
-    // (part of public API).
-    if (!autograd_meta_) impl::GetAutogradMetaFactory()->undefined_tensor();
-    return autograd_meta_->grad();
-  } else {
-    AT_ERROR("grad is not implemented for Tensor");
-  }
+  // Yes, I know this looks really weird.  But I don't really have a choice as
+  // long as this function returns a const reference to Tensor.  I'm not
+  // really sure how I would have designed this API differently, but it
+  // is not so easy to fix right now because the mutable counterpart of
+  // this function must keep working so that "x.grad() = ..." keeps working
+  // (part of public API).
+  if (!autograd_meta_) impl::GetAutogradMetaFactory()->undefined_tensor();
```

Comment:
I think I'm missing something here. I can't find an `undefined_tensor()` method on the factory.

Reply:
Factory, not interface.

```
struct C10_API AutogradMetaFactory {
  virtual ~AutogradMetaFactory() = default;
  virtual std::unique_ptr<AutogradMetaInterface> make() const = 0;
  // This method is the dumbest method.  But I don't have access
  // to Tensor (not TensorImpl) which is undefined in this header.
  virtual const at::Tensor& undefined_tensor() const = 0;
};
```

Reply:
Why do you call this function and do nothing with the result?

Reply:
oooh, I'm supposed to return it. good catch